src/Service/EndpointService.php

```diff
@@ -282,6 +281,93 @@ public function getEndpoint(): Endpoint
 
     }//end getEndpoint()
 
+    /**
+     * Parse headers from multipart/form-data requests on put, as PHP does not parse them itself
+     *
+     * @param string $rawHeaders The raw headers.
+     *
+     * @return array The parsed headers.
+     */
+    private function parsePutHeaders(string $rawHeaders): array
+    {
+        $rawHeaders = explode("\r\n", $rawHeaders);
+        $headers    = [];
+        foreach ($rawHeaders as $header) {
+            list($name, $value)         = explode(':', $header);
+            $headers[strtolower($name)] = ltrim($value, ' ');
+        }
+
+        return $headers;
+
+    }//end parsePutHeaders()
+
+    /**
+     * Retrieves data from input on PUT requests where multipart/form-data is not natively supported by PHP
+     *
+     * @return array The data from the PUT request parsed into an array
+     *
+     * @throws \Safe\Exceptions\FilesystemException
+     */
+    private function getPutData(): array
+    {
+        // Fetch content and determine boundary
+        $raw_data = \Safe\file_get_contents('php://input');
+        $boundary = substr($raw_data, 0, strpos($raw_data, "\r\n"));
+
+        // Fetch each part
+        $parts = array_slice(explode($boundary, $raw_data), 1);
+        $data  = [];
+
+        foreach ($parts as $part) {
+            // If this is the last part, break
+            if ($part == "--\r\n") {
+                break;
+            }
+
+            // Separate content from headers
+            $part                    = ltrim($part, "\r\n");
+            list($rawHeaders, $body) = explode("\r\n\r\n", $part, 2);
+
+            // Parse the headers list
+            $this->parsePutHeaders($rawHeaders);
+
+            // Parse the Content-Disposition to get the field name, etc.
+            if (isset($headers['content-disposition']) === false) {
+                continue;
+            }
+
+            $filename = null;
+            preg_match(
+                '/^(.+); *name="([^"]+)"(; *filename="([^"]+)")?/',
+                $headers['content-disposition'],
+                $matches
+            ) ?: preg_match(
+                '/^(.+); *name=([-+.\w]+)(; *filename=([-+.\w]+))?/',
+                $headers['content-disposition'],
+                $matches
+            );
+
+            list(, $type, $name)             = $matches;
+            isset($matches[4]) and $filename = $matches[4];
+
+            // handle your fields here
+            switch ($name) {
+                // this is a file upload
+            case 'userfile':
+                file_put_contents($filename, $body);
```

Comment:
Can you end up here if $name = 'userfile' but $filename isn't set?

because of:
`list(, $type, $name)             = $matches;`
`isset($matches[4]) and $filename = $matches[4];`